### DISCUSSION TITLE
data dictionary in _arrange_data function

### DISCUSSION THREAD

**Discussion by ps96068** [jowilf/starlette-admin#606] [Q&A] [answered]
dear author,

In the process of implementing the library in my project, I encountered the following problem:

I have the User model, this model has several fields, the problem is in the following (code from the User model):

```
company_id: Mapped[int | None] = mapped_column(ForeignKey("companies.id"), nullable=True)
status: Mapped[UserStatus] = mapped_column(Enum(UserStatus), nullable=False)
```

Here is the enum class code from the model:

```
class UserStatus(enum.Enum):
    user = "user"
    anonim = "anonim"
    company = "company"
```

For this model I created UserModelView and set the data fields: exclude_from_create=True.
if we look in the code starlette-admin -> fields.BaseField, it is written as:
"exclude_from_create: Control field visibility in create page"

from this, I concluded that the field is just not visible!
For UserModelView I created the "create" method in which I update the "data" dictionary:

```
data.update(
            {
                "company_id": 1,
                "password_hash": get_hashed_password(data["password_hash"]),
                "status": UserStatus.company.value,
            }
        )
```
in the database i have company with id=1.
This dictionary is sent to the ModelView._arrange_data function, this function returns the "arranged_data" dictionary in which these manually added fields are missing.

The first idea that came to my mind is to rewrite this method in UserModelView, but maybe there is another method.
Can you give me some advice?

**Comment by ps96068** [comment]
> dear author,
> 
> In the process of implementing the library in my project, I encountered the following problem:
> 
> I have the User model, this model has several fields, the problem is in the following (code from the User model):
> 
> ```
> company_id: Mapped[int | None] = mapped_column(ForeignKey("companies.id"), nullable=True)
> status: Mapped[UserStatus] = mapped_column(Enum(UserStatus), nullable=False)
> ```
> 
> Here is the enum class code from the model:
> 
> ```
> class UserStatus(enum.Enum):
>     user = "user"
>     anonim = "anonim"
>     company = "company"
> ```
> 
> For this model I created UserModelView and set the data fields: exclude_from_create=True. if we look in the code starlette-admin -> fields.BaseField, it is written as: "exclude_from_create: Control field visibility in create page"
> 
> from this, I concluded that the field is just not visible! For UserModelView I created the "create" method in which I update the "data" dictionary:
> 
> ```
> data.update(
>             {
>                 "company_id": 1,
>                 "password_hash": get_hashed_password(data["password_hash"]),
>                 "status": UserStatus.company.value,
>             }
>         )
> ```
> 
> in the database i have company with id=1. This dictionary is sent to the ModelView._arrange_data function, this function returns the "arranged_data" dictionary in which these manually added fields are missing.
> 
> The first idea that came to my mind is to rewrite this method in UserModelView, but maybe there is another method. Can you give me some advice?

I found a solution, but I can't really tell how good it is.
the solution is:

I redefined in my class, UserModelView, the "_populate_obj" method, at the end of the function where the "obj" object is already created, I manually add all the fields I need in the following way:

```

class UserModelView(ModelView):
	fields = [....]
	........
	
	async def _populate_obj(
        self,
        request: Request,
        obj: Any,
        data: Dict[str, Any],
        is_edit: bool = False,
    ) -> Any:
		.....
		.....
		
		setattr(obj, 'company_id', 1)
		
		return obj

```

dear author, what do you say?

**Reply by jowilf** [reply, marked as answer]
I think what you are looking for is https://jowilf.github.io/starlette-admin/user-guide/configurations/modelview/#hooks

**Reply by ps96068** [reply]
thanks
